{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create SRA submission sheet\n",
    "This Python Jupyter notebook creates a BioSample submission sheet for the Sequence Read Archive:\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import glob\n",
    "import itertools\n",
    "import os\n",
    "\n",
    "import natsort\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display full columns of data frames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the configuration for the uploads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('upload_config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the Illumina runs to submit:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Illumina runs from barcode_runs_to_upload.csv\n",
      "Here are the first few entries in the Illumina runs:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>experiment</th>\n",
       "      <th>target</th>\n",
       "      <th>library</th>\n",
       "      <th>antibody</th>\n",
       "      <th>concentration</th>\n",
       "      <th>sort_bin</th>\n",
       "      <th>selection</th>\n",
       "      <th>sample</th>\n",
       "      <th>experiment_type</th>\n",
       "      <th>number_cells</th>\n",
       "      <th>frac_escape</th>\n",
       "      <th>R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>220816</td>\n",
       "      <td>exptREF</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib12</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ref</td>\n",
       "      <td>reference</td>\n",
       "      <td>exptREF-none-0-ref</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref1_S70_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref2_S71_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref3_S72_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref4_S73_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>220816</td>\n",
       "      <td>exptREF</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib13</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ref</td>\n",
       "      <td>reference</td>\n",
       "      <td>exptREF-none-0-ref</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref1_S74_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref2_S75_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref3_S76_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref4_S77_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>220816</td>\n",
       "      <td>expt8</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib12</td>\n",
       "      <td>C68_88</td>\n",
       "      <td>384</td>\n",
       "      <td>abneg</td>\n",
       "      <td>escape</td>\n",
       "      <td>expt8-C68_88-384-abneg</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>204245.0</td>\n",
       "      <td>0.056</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/expt8_lib1_S15_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>220816</td>\n",
       "      <td>expt8</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib13</td>\n",
       "      <td>C68_88</td>\n",
       "      <td>384</td>\n",
       "      <td>abneg</td>\n",
       "      <td>escape</td>\n",
       "      <td>expt8-C68_88-384-abneg</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>210346.0</td>\n",
       "      <td>0.059</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/expt8_lib2_S16_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230714</td>\n",
       "      <td>exptREF2</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib12</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ref</td>\n",
       "      <td>reference</td>\n",
       "      <td>exptREF2-none-0-ref</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-1_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-2_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-3_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-4_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     date experiment      target library antibody  concentration sort_bin  \\\n",
       "0  220816    exptREF  Wuhan_Hu_1   lib12     none              0      ref   \n",
       "1  220816    exptREF  Wuhan_Hu_1   lib13     none              0      ref   \n",
       "2  220816      expt8  Wuhan_Hu_1   lib12   C68_88            384    abneg   \n",
       "3  220816      expt8  Wuhan_Hu_1   lib13   C68_88            384    abneg   \n",
       "4  230714   exptREF2  Wuhan_Hu_1   lib12     none              0      ref   \n",
       "\n",
       "   selection                  sample experiment_type  number_cells  \\\n",
       "0  reference      exptREF-none-0-ref    ab_selection           NaN   \n",
       "1  reference      exptREF-none-0-ref    ab_selection           NaN   \n",
       "2     escape  expt8-C68_88-384-abneg    ab_selection      204245.0   \n",
       "3     escape  expt8-C68_88-384-abneg    ab_selection      210346.0   \n",
       "4  reference     exptREF2-none-0-ref    ab_selection           NaN   \n",
       "\n",
       "   frac_escape  \\\n",
       "0          NaN   \n",
       "1          NaN   \n",
       "2        0.056   \n",
       "3        0.059   \n",
       "4          NaN   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   R1  \n",
       "0                                                                              /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref1_S70_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref2_S71_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref3_S72_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref4_S73_R1_001.fastq.gz  \n",
       "1                                                                              /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref1_S74_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref2_S75_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref3_S76_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref4_S77_R1_001.fastq.gz  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/expt8_lib1_S15_R1_001.fastq.gz  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/expt8_lib2_S16_R1_001.fastq.gz  \n",
       "4  /uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-1_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-2_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-3_R1_001.fastq.gz; /uufs/chpc.utah.edu/common/home/starr-group1/sequencing/TNS/2023/230925_Azenta-pools/230714_Overbaugh-DMS-batch2/2023-07-14-lib12-pre-4_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Reading Illumina runs from {config['barcode_runs']}\")\n",
    "\n",
    "illumina_runs = pd.read_csv(config['barcode_runs'])\n",
    "\n",
    "print('Here are the first few entries in the Illumina runs:')\n",
    "illumina_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for the barcode runs, initially in \"tidy\" format with one FASTQ file per row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in tidy format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref1_S70_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref1_S70_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref2_S71_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref2_S71_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref3_S72_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref3_S72_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref4_S73_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref4_S73_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib13_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref1_S74_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref1_S74_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession                library_ID  \\\n",
       "0        SAMN40905401  lib12_exptREF-none-0-ref   \n",
       "1        SAMN40905401  lib12_exptREF-none-0-ref   \n",
       "2        SAMN40905401  lib12_exptREF-none-0-ref   \n",
       "3        SAMN40905401  lib12_exptREF-none-0-ref   \n",
       "4        SAMN40905401  lib13_exptREF-none-0-ref   \n",
       "\n",
       "                                                                                                                 title  \\\n",
       "0  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "1  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "2  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "3  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "4  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "\n",
       "  library_strategy library_source library_selection library_layout  platform  \\\n",
       "0         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "  instrument_model                 design_description filetype  \\\n",
       "0     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "1     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "2     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "3     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "4     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                                                                                                      filename_fullpath  \\\n",
       "0  /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref1_S70_R1_001.fastq.gz   \n",
       "1  /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref2_S71_R1_001.fastq.gz   \n",
       "2  /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref3_S72_R1_001.fastq.gz   \n",
       "3  /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib1_ref4_S73_R1_001.fastq.gz   \n",
       "4  /uufs/chpc.utah.edu/common/home/u6042467/starr-group1/sequencing/TNS/2022/220816_Overbaugh-mAb-MAP/WH1_lib2_ref1_S74_R1_001.fastq.gz   \n",
       "\n",
       "                            filename  \n",
       "0  WH1_lib1_ref1_S70_R1_001.fastq.gz  \n",
       "1  WH1_lib1_ref2_S71_R1_001.fastq.gz  \n",
       "2  WH1_lib1_ref3_S72_R1_001.fastq.gz  \n",
       "3  WH1_lib1_ref4_S73_R1_001.fastq.gz  \n",
       "4  WH1_lib2_ref1_S74_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_tidy = (\n",
    "    illumina_runs\n",
    "    .assign(\n",
    "        biosample_accession=lambda x: config['biosample_accession'],\n",
    "        library_ID=lambda x: x['library'] + '_' + x['sample'],\n",
    "        title=lambda x: 'Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD ' + x['target'],\n",
    "        library_strategy='AMPLICON',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='PCR',\n",
    "        library_layout='single',\n",
    "        platform='ILLUMINA',\n",
    "        instrument_model='NextSeq 2000',\n",
    "        design_description='PCR of barcodes from RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['R1'].str.split(';')\n",
    "                                    .map(lambda flist: list(itertools.chain.from_iterable(glob.glob(f.strip()) for f in flist))),       \n",
    "        )\n",
    "    .explode('filename_fullpath')\n",
    "    .assign(filename_fullpath=lambda x: x['filename_fullpath'].str.strip(),\n",
    "            filename=lambda x: x['filename_fullpath'].map(os.path.basename))\n",
    "    .drop(columns=illumina_runs.columns)\n",
    "    .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "assert submissions_tidy['filename_fullpath'].map(os.path.isfile).all()\n",
    "\n",
    "print('Here are the first few submission entries in tidy format:')\n",
    "submissions_tidy.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the actual submission, we need a \"wide\" data frame that for each unique `sample_name` / `library_ID` gives all of the files each in different columns.\n",
    "These should be files without the full path.\n",
    "\n",
    "First, look at how many files there are for each sample / library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>n_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_exptREF3-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib13_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_exptREF3-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib13_exptREF2-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_exptREF2-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_exptREF2-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF2-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_exptREF-none-0-ref</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt6-C68_348-79-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt5-C68_201-50-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt4-C68_327-259-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt3-C68_203-107-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt2-C68_200-312-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt2-C68_10-77-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt16-C68_185-110-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt15-C68_183-180-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt13-C68_121-95-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt12-C68_61-98-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt1-C68_175-93-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib25_expt3-C68_83-34-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_expt7-C68_239-73-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_expt8-C68_88-384-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt5-C68_201-50-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt4-C68_327-259-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt3-C68_83-34-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt3-C68_203-107-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt2-C68_200-312-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt2-C68_10-77-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt16-C68_185-110-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt15-C68_183-180-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt13-C68_121-95-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt12-C68_61-98-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt1-C68_175-93-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib13_expt8-C68_88-384-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib13_expt7-C68_239-73-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib24_expt6-C68_348-79-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_accession                      library_ID  n_files\n",
       "0         SAMN40905401       lib25_exptREF3-none-0-ref        4\n",
       "1         SAMN40905401        lib13_exptREF-none-0-ref        4\n",
       "2         SAMN40905401        lib24_exptREF-none-0-ref        4\n",
       "3         SAMN40905401       lib24_exptREF3-none-0-ref        4\n",
       "4         SAMN40905401       lib13_exptREF2-none-0-ref        4\n",
       "5         SAMN40905401       lib24_exptREF2-none-0-ref        4\n",
       "6         SAMN40905401       lib25_exptREF2-none-0-ref        4\n",
       "7         SAMN40905401       lib12_exptREF2-none-0-ref        4\n",
       "8         SAMN40905401        lib12_exptREF-none-0-ref        4\n",
       "9         SAMN40905401        lib25_exptREF-none-0-ref        2\n",
       "10        SAMN40905401    lib25_expt6-C68_348-79-abneg        1\n",
       "11        SAMN40905401    lib25_expt5-C68_201-50-abneg        1\n",
       "12        SAMN40905401   lib25_expt4-C68_327-259-abneg        1\n",
       "13        SAMN40905401   lib25_expt3-C68_203-107-abneg        1\n",
       "14        SAMN40905401   lib25_expt2-C68_200-312-abneg        1\n",
       "15        SAMN40905401     lib25_expt2-C68_10-77-abneg        1\n",
       "16        SAMN40905401  lib25_expt16-C68_185-110-abneg        1\n",
       "17        SAMN40905401  lib25_expt15-C68_183-180-abneg        1\n",
       "18        SAMN40905401   lib25_expt13-C68_121-95-abneg        1\n",
       "19        SAMN40905401    lib25_expt12-C68_61-98-abneg        1\n",
       "20        SAMN40905401    lib25_expt1-C68_175-93-abneg        1\n",
       "21        SAMN40905401     lib25_expt3-C68_83-34-abneg        1\n",
       "22        SAMN40905401    lib12_expt7-C68_239-73-abneg        1\n",
       "23        SAMN40905401    lib12_expt8-C68_88-384-abneg        1\n",
       "24        SAMN40905401    lib24_expt5-C68_201-50-abneg        1\n",
       "25        SAMN40905401   lib24_expt4-C68_327-259-abneg        1\n",
       "26        SAMN40905401     lib24_expt3-C68_83-34-abneg        1\n",
       "27        SAMN40905401   lib24_expt3-C68_203-107-abneg        1\n",
       "28        SAMN40905401   lib24_expt2-C68_200-312-abneg        1\n",
       "29        SAMN40905401     lib24_expt2-C68_10-77-abneg        1\n",
       "30        SAMN40905401  lib24_expt16-C68_185-110-abneg        1\n",
       "31        SAMN40905401  lib24_expt15-C68_183-180-abneg        1\n",
       "32        SAMN40905401   lib24_expt13-C68_121-95-abneg        1\n",
       "33        SAMN40905401    lib24_expt12-C68_61-98-abneg        1\n",
       "34        SAMN40905401    lib24_expt1-C68_175-93-abneg        1\n",
       "35        SAMN40905401    lib13_expt8-C68_88-384-abneg        1\n",
       "36        SAMN40905401    lib13_expt7-C68_239-73-abneg        1\n",
       "37        SAMN40905401    lib24_expt6-C68_348-79-abneg        1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(submissions_tidy\n",
    " .groupby(['biosample_accession', 'library_ID'])\n",
    " .aggregate(n_files=pd.NamedAgg('filename_fullpath', 'count'))\n",
    " .sort_values('n_files', ascending=False)\n",
    " .reset_index()\n",
    " )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now make the wide submission data frame.\n",
    "Note we keep only the filename column with the path lacking the full directory information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in wide format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>filename3</th>\n",
       "      <th>filename4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>WH1_lib1_ref1_S70_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref2_S71_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref3_S72_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref4_S73_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib13_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>WH1_lib2_ref1_S74_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref2_S75_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref3_S76_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref4_S77_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_expt8-C68_88-384-abneg</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>expt8_lib1_S15_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib13_expt8-C68_88-384-abneg</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>expt8_lib2_S16_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN40905401</td>\n",
       "      <td>lib12_exptREF2-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>2023-07-14-lib12-pre-1_R1_001.fastq.gz</td>\n",
       "      <td>2023-07-14-lib12-pre-2_R1_001.fastq.gz</td>\n",
       "      <td>2023-07-14-lib12-pre-3_R1_001.fastq.gz</td>\n",
       "      <td>2023-07-14-lib12-pre-4_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession                    library_ID  \\\n",
       "0        SAMN40905401      lib12_exptREF-none-0-ref   \n",
       "1        SAMN40905401      lib13_exptREF-none-0-ref   \n",
       "2        SAMN40905401  lib12_expt8-C68_88-384-abneg   \n",
       "3        SAMN40905401  lib13_expt8-C68_88-384-abneg   \n",
       "4        SAMN40905401     lib12_exptREF2-none-0-ref   \n",
       "\n",
       "                                                                                                                 title  \\\n",
       "0  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "1  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "2  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "3  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "4  Illumina barcode sequencing from C68 donor mAb (v1) mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "\n",
       "  library_strategy library_source library_selection library_layout  platform  \\\n",
       "0         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "  instrument_model                 design_description filetype  \\\n",
       "0     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "1     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "2     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "3     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "4     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                 filename  \\\n",
       "0       WH1_lib1_ref1_S70_R1_001.fastq.gz   \n",
       "1       WH1_lib2_ref1_S74_R1_001.fastq.gz   \n",
       "2          expt8_lib1_S15_R1_001.fastq.gz   \n",
       "3          expt8_lib2_S16_R1_001.fastq.gz   \n",
       "4  2023-07-14-lib12-pre-1_R1_001.fastq.gz   \n",
       "\n",
       "                                filename2  \\\n",
       "0       WH1_lib1_ref2_S71_R1_001.fastq.gz   \n",
       "1       WH1_lib2_ref2_S75_R1_001.fastq.gz   \n",
       "2                                     NaN   \n",
       "3                                     NaN   \n",
       "4  2023-07-14-lib12-pre-2_R1_001.fastq.gz   \n",
       "\n",
       "                                filename3  \\\n",
       "0       WH1_lib1_ref3_S72_R1_001.fastq.gz   \n",
       "1       WH1_lib2_ref3_S76_R1_001.fastq.gz   \n",
       "2                                     NaN   \n",
       "3                                     NaN   \n",
       "4  2023-07-14-lib12-pre-3_R1_001.fastq.gz   \n",
       "\n",
       "                                filename4  \n",
       "0       WH1_lib1_ref4_S73_R1_001.fastq.gz  \n",
       "1       WH1_lib2_ref4_S77_R1_001.fastq.gz  \n",
       "2                                     NaN  \n",
       "3                                     NaN  \n",
       "4  2023-07-14-lib12-pre-4_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .assign(\n",
    "        filename_count=lambda x: x.groupby(['biosample_accession', 'library_ID'])['filename'].cumcount() + 1,\n",
    "        filename_col=lambda x: 'filename' + x['filename_count'].map(lambda c: str(c) if c > 1 else '')\n",
    "        )\n",
    "    .pivot(\n",
    "        index='library_ID',\n",
    "        columns='filename_col',\n",
    "        values='filename',\n",
    "        )\n",
    "    )\n",
    "\n",
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .drop(columns=['filename_fullpath', 'filename'])\n",
    "    .drop_duplicates()\n",
    "    .merge(submissions_wide[natsort.natsorted(submissions_wide.columns)],\n",
    "           on='library_ID',\n",
    "           validate='one_to_one',\n",
    "           )\n",
    "    )\n",
    "\n",
    "print('Here are the first few submission entries in wide format:')\n",
    "submissions_wide.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now write the wide submissions data frame to a `*.tsv` file that can be used for uploading the SRA submission website:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "submissions_spreadsheet = 'SRA_submission_spreadsheet.tsv'\n",
    "\n",
    "submissions_wide.to_csv(submissions_spreadsheet, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also want to write a file with all of the FASTQ files that are in the submission spreadsheet:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing list of all the FASTQ files to upload to FASTQs_to_upload.csv\n"
     ]
    }
   ],
   "source": [
    "fastq_file_list = 'FASTQs_to_upload.csv'\n",
    "\n",
    "print(f\"Writing list of all the FASTQ files to upload to {fastq_file_list}\")\n",
    "\n",
    "(submissions_tidy\n",
    " [['filename_fullpath', 'filename']]\n",
    " .to_csv(fastq_file_list, index=False)\n",
    " )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
